### PR TITLE
Add environments.md

### PR DESCRIPTION
**Pull request by ellismg**
Using this as an experiment for discussing an updated design for environments.

### CODE REVIEW

**Comment by ellismg** [comment]
This document captures a description of how environments are used today and proposes a new design which de-emphasizes the environment name. The goal of this work is to explore a possible design where the concept of an environment does not need to be introduced early on when someone comes to `azd` and instead can learn about environments (and the fact that they can have more than one) until they get to a point where doing so would be valuable.

File: cli/azd/design-notes/environments.md [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+`current-principal-id`: The value should be the id of the authenticated principal.
+
+Configuration values (including the environment name and id) are **not** stored in the `.env` file.
+
+Deployment outputs are also stored in the `config.json` file under the "outputs" key (a sibling of "properties") and updated on each deployment. They are written to the `.env` file as before, as well.
+
+The intermediate `.parameters.json` files are no longer written in the `.azure\<environment-name>` directory instead they are written in a temporary directory (and when we stop depending on `az` for deployments will not be persisted to disk)
+
+When creating names for deployments, we name them using the combination of the environment name and ID (e.g. `default-1539e0b8-ac32-440c-ae1f-d131512563dd`), if this would generate a deployment name which is too long, we truncate the environment name but keep the full id). This ensures the deployment names don't clash between two unrelated projects even if they have the same environment name (which will now be common, since we have a default value and no longer force the user into picking a new environment).
```
[the remark this answers is not in the document]
Comment (ellismg) [reply cli/azd/design-notes/environments.md:188]: > I'm thinking about what the resource group/resources will look like when they go to the portal or look at things in the VS Code siderail. I'm imagining that if they have a couple projects all starting with "default-..." then the names are pretty hard to parse (because they'd be relying on remembering/knowing the random ID to figure out which resource group belongs to which project etc.)

This is subtle, but the name of the deployment has no impact on the resources provisioned by that deployment.  It's simply a named object that is used to hold information about the deployment that is progressing.

The name ends up being visible to users if they look at the "deployments" tab of their subscription (or when we target resource groups, looking at the "deployments" tab of the resource group) and in the URL we print during a `provision` to view progress on the Azure portal, but otherwise is not something developers would think about day to day.

If we adopt this model, our templates will still take a `name` parameter, and that name will be used to create the resource group and the user has full control over this name (just as they have full control over any other parameter of their deployment)

Having said that, I can see the value of including the project name in this string - If we adopt `<project-name>-<env-name>-<env-id>` we'll end up with names that may be too longer for a deployment (the maximum length is 64 characters, the UUID takes 36 characters, "default" is 7 characters and the two hypens between these tokens and project names are usually longer strings right now like `todo-nodejs-mongo-aca` so we'd have to do some level of shortening.  `<project-name>-<environment-id>` is nice because it is shorter and less likely to have to be truncated but it does make the names a bit harder to interpret.

> We'd have to prompt for the project name or just infer it from the template/directory name.

There is actually a `name` property in azure.yaml today which we can use as the project name (and if you run `azd init` without a template the name defaults to the name of the directory you are initializing into)

File: cli/azd/design-notes/environments.md [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+
+```json
+{
+  "id": "9fdfa455-2131-4faf-b2e3-8380aba97c96",
+  "name": "default",
+  "infra": {
+    "main": { 
+      "target": {
+        "type": "arm",
+        "subscriptionId": "faa080af-c1d8-40ad-9cce-e1a450ca5b57",
```
[the remark this answers is not in the document]
Comment (ellismg) [reply cli/azd/design-notes/environments.md:202]: We would have to design a gesture, but I agree that it important for us to do so.  I think we will need to think through this more as we start to think about different deployment backends.

We've imagined an `azd infra show` type command that lists information about your application's infrastructure (key resources, portal links, etc) - this could be a good place to surface this.

The way this is knowable today is by running `azd env get-values` and just knowing that the value of `AZURE_SUBSCRIPTION_ID` is what the deployment system targets.

File: cli/azd/design-notes/environments.md [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+
+This document outlines what an environment is in `azd` today and outlines a change in model to remove the need to introduce the concept on the critical path for a user getting started with `azd` to manage their application.
+
+## Current Design
+
+The concept on an environment has been present in `azd` for a long while, but the way it has been used has changed over time. In addition, the ties between an environment name and provisioned resources has changed over time.
+
+Every command in `azd` today runs in the context of an environment. The name of the environment can be provided by the `-e` flag which is command to all `azd` commands and when not set, the *default environment* is used. The name of the currently selected environment is stored in `.azure/config.json` file, under the `defaultEnvironment`. The *default environment* can set with `azd env select [environment-name]`. If no environment name is passed and no default environment has been selected then the user is prompted to create a new environment before the operation can proceed.
+
+During environment creation the user is prompted for the following pieces of information:
```
[the remark this answers is not in the document]
Comment (ellismg) [reply cli/azd/design-notes/environments.md:18]: Yes - thanks - I will try to make this clearer.

**Review by jongio — APPROVED** [review]

File: cli/azd/design-notes/environments.md [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+
+- `AZURE_ENV_NAME`: The environment name.
+- `AZURE_LOCATION`: The short location name (e.g. `eastus2`)
+- `AZURE_SUBSCRIPTION_ID`: The id of the selected subscription (e.g. `faa080af-c1d8-40ad-9cce-e1a450ca5b57`)
+
+In addition, the following value is added without any user interaction:
+
+- `AZURE_PRINCIPAL_ID`: The id of the current principal authenticated to the `az` CLI.
+
+All of these values must be present in an environment for it to be considered initialized.  If any are missing (for example, by editing the .env file) the user will be prompted the next time `azd` runs to configure these values. Note that this will require the user be logged into the `az` CLI since configuring these values requires the CLI (to list locations, subscriptions and the object id of the current principal).
```
[the remark this answers is not in the document]
Comment (ellismg) [reply cli/azd/design-notes/environments.md:38]: Yes - this is true today and enforced by [`ensureEnvironmentInitialized`](https://github.com/Azure/azure-dev/blob/77ab01c3bcf40d12852d543506ecbb935a779f81/cli/azd/cmd/util.go#L168-L256).

I agree that it is not /technically/ required - you don't need this piece of information to create an azure deployment and our bicep templates are written such that if this value was unset things would deploy successfully. It would break our end to end experience where we run the app locally but connect to the cosmos db instance in the cloud (and other things like KeyVault where we expect to leverage managed identity and create access rules as part of infrastructure deployment).

I suspect this is the reason we ended up treating it as something that should always be present (the fact that the value is fixed and doesn't reflect the "current" principal but rather the principal we had when we initialized the environment is likely a bug in the current implementation we should fix if we don't adopt this design.

File: cli/azd/design-notes/environments.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (ellismg) [reply cli/azd/design-notes/environments.md:188]: (more of an observation to myself) We may not need to include the entire id in the deployment name, but we should make it possible to collect all deployments for an environment by its ID somehow (I believe you can associate a tag with the deployment object, but this may not be possible with the CLI)

I suspect we will also want to consider adding a timestamp to the name or some other value that changes per deployment - not having a separate deployment object per deployment seems like it will be frustrating long term (we already see this a bit with some of the "get deployments operations to show progress" stuff we do since we can end up seeing the previous deployment operations deployment since it's URI and the new deployment URI end up being the same.  If we had separate names this would no longer be an issue.

[the remark this answers is not in the document]
Comment (ellismg) [reply cli/azd/design-notes/environments.md]: Adjusted based on the discussion here.

File: cli/azd/design-notes/environments.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (ellismg) [reply cli/azd/design-notes/environments.md:232]: Added some comments about how to manage this as part of my last pass over the doc.

**Review by puicchan — APPROVED** [review]

File: cli/azd/design-notes/environments.md [hunk 5]
```diff
@@ -0,0 +1,271 @@
...
+
+We'll have a services object in the context which will have keys for each service (mirroring `azure.yaml`) and it will contain an `imageId` property.
+
+> How does this impact `azd env refresh` - What does refreshing an environment mean?  
+
+`env refresh` fetches the latest deployment object for the environment and updates `properties` and `outputs` sections of the `config.json` file for the `main` module. As a consequence of changing `config.json`, the `.env` and `.parameters.json` files are also regenerated.  (*note: I would honestly prefer if we just got rid of this command for now, I don't think we have a great understanding of what this tries to solve and the value it provides today is quite limited*)
```
Comment (jongio) [inline cli/azd/design-notes/environments.md:269]: Do you mean remove `env refresh`?  This command is helpful when I want to share an environment with someone without having them provision it.  I can tell them to run `azd env refresh -e envname`

File: cli/azd/design-notes/environments.md [hunk 6]
```diff
@@ -0,0 +1,271 @@
...
+
+Note that the target **does not** contain information about a deployment name. Today the name we use for the deployment matches the environment name. We will change this to instead include the *project name* (stored in the name property of `azure.yaml` and the *environment id* to ensure uniqueness. We should also include some per deployment entropy (or perhaps just a timestamp) such that each deployment gets its own deployment object. This should help with some problems we face today where re-using the same deployment object can confuse our status reporting logic, and provides a better view of deployment history.  
+
+Since deployment names are limited to 64 characters, we will need someway to truncate names if we hit this limit.  We should bias towards keeping as much entropy as possible in the strings to ensure uniqueness at a given scope. We should tag these deployment objects with the full environment ID for querying later.
+
+## User Impacting Changes
+
+- We no longer add `AZURE_ENV_NAME`, `AZURE_LOCATION`, `AZURE_SUBSCRIPTION_ID` or `AZURE_PRINCIPAL_ID` to the `.env` file.
```
Comment (jongio) [inline cli/azd/design-notes/environments.md:250]: Please verify that these values are not used in any of the templates via .env.  Also, @karolz-ms loads values and may use them for ide integration.

File: cli/azd/design-notes/environments.md [hunk 7]
```diff
@@ -0,0 +1,271 @@
...
+- We no longer support environment variable substitution in the `.parameters.json` file in the `infra` folder, it is used just to configure static values.
+- In our templates we remove the `.parameters.json` files since they are no longer used (all values are user configured during first deployment). (*note: We can choose to keep these files if we think their existence provides value in our templates as a learning tool*)
```
Comment (jongio) [inline cli/azd/design-notes/environments.md:252]: If we keep them, then the values a user enters into these files should be honored while doing a deployment.

File: cli/azd/design-notes/environments.md [hunk 8]
```diff
@@ -0,0 +1,271 @@
...
+- We prompt and validate deployment parameters using the metadata applied in the bicep template.
+
+We can figure out how to rolls these changes out in a smooth enough way (for example, we could take the existence of a environment reference like `${FOO}` in a `.parameters.json` to mean we were on the "old" plan and `azd` can warn and do something sensible (or point folks at an article on how to upgrade).
+
+## Open Questions
+
+> How do we express that a deployment parameter should correspond to the environment id or environment name. Do we introduce types like `azd-env-id` and `azd-env-name` for our metadata on parameters? Should we instead have something like `azd-context` and that passes in a JSON object with well known keys like `id` and `name`?  Is there value in making it easy to flow the environment name (which now will often just be the string "default" into the infrastructure? Why would you use it vs `azd-env-id`?)
```
Comment (jongio) [inline cli/azd/design-notes/environments.md:259]: How do I **easily** associate a ANY parameter with an environment variable?  Right now that is done via parameters.json and env var subs and is easy to see what is happening and explain how it happens.  We shouldn't lose that simplicity.

File: cli/azd/design-notes/environments.md [hunk 9]
```diff
@@ -0,0 +1,271 @@
...
+
+Note that the target **does not** contain information about a deployment name. Today the name we use for the deployment matches the environment name. We will change this to instead include the *project name* (stored in the name property of `azure.yaml` and the *environment id* to ensure uniqueness. We should also include some per deployment entropy (or perhaps just a timestamp) such that each deployment gets its own deployment object. This should help with some problems we face today where re-using the same deployment object can confuse our status reporting logic, and provides a better view of deployment history.  
+
+Since deployment names are limited to 64 characters, we will need someway to truncate names if we hit this limit.  We should bias towards keeping as much entropy as possible in the strings to ensure uniqueness at a given scope. We should tag these deployment objects with the full environment ID for querying later.
+
+## User Impacting Changes
```
Comment (jongio) [inline cli/azd/design-notes/environments.md:248]: The bicep parameter files with the resulting values should be outputted to the env directory so the user can then later use them without azd if they'd like to.  And they can use it as a debugging tool to see what the values are being passed.

Comment (jongio) [inline cli/azd/design-notes/environments.md:248]: The .env file should also be outputted so the user can find it and use it in their application.

File: cli/azd/design-notes/environments.md [hunk 10]
```diff
@@ -0,0 +1,271 @@
...
+
+This system allows you to create a multiple environments, each with a different name, and provision a unique copy of the infrastructure for the application.
+
+## Current Challenges
+
+The major issue we face with the current design is that we've introduced a new concept, an environment, and it is on the critical path of every `azd` invocation. We need to ensure this concept is simple to understand and worthwhile and clearly articulate how it relates to the the rest of the system (and ideally existing Azure concepts). It's not always clear how settings in the environment impact infrastructure or the rest of `azd` and what's based on convention and what is required by the tooling. There are also "tiers" of configuration values now - `AZURE_ENV_NAME`, `AZURE_LOCATION`, `AZURE_PRINCIPAL_ID` and `AZURE_SUBSCRIPTION_ID` are very different from other values. We also have a tension on how to describe what an environment name does. Our templates today use it to influence the names of some resources (usually the root resource group that is created) but that's simply a convention we've adopted and so we are wary of adding that as "help text" for an environment name.
```
Comment (jongio) [inline cli/azd/design-notes/environments.md:107]: What do you mean by 'very different'?

File: cli/azd/design-notes/environments.md [hunk 11]
```diff
@@ -0,0 +1,271 @@
...
+
+The value of environments doesn't pay off until you want to start creating multiple copies of the infrastructure that backs your application, but the complexity cost is incurred as a soon as a user walks up to `azd`.
+
+I think that with a few small tweaks to our model we can make it much easier to understand and describe - while at the same time allowing us to defer introduction of the concept until a user actually cares about the value an environment provides.
+
+## Proposed Model
+
+### Overview
+
+As before, we still have the concept of an environment in our system.  All `azd` commands still accept an `-e` parameter to select what environment is used, and `azd env select` can be used to select the default environment. However, if no environment is selected, `azd` creates a new environment with the name `default`. In addition to a name every environment gets a unique id (a v4 UUID) assigned to it. These values are recorded in `.azure/<environment-name>/config.json`:
```
Comment (jongio) [inline cli/azd/design-notes/environments.md:117]: I'm concerned about creating an environment with a name that the user didn't select. I think that is further abstracting the concept of environment rather than embracing it. 

Instead we should guide them through the process, explain what things means, and help them make informed decisions.

File: cli/azd/design-notes/environments.md [hunk 12]
```diff
@@ -0,0 +1,271 @@
...
+    "main": { 
+      "properties": {
+         // ... the object that is set as the value of the "properties" key of a deploymentParameters file
+      }
+    }
+  }
+}
+```
+
+To save ourselves the trouble of designing our own format, we reuse the schema of a `deploymentParamters` file. One nice feature here is that system provides enough flexibility to express references to secrets in Azure Key Vault, which will be a useful feature for us down the line.
```
Comment (jongio) [inline cli/azd/design-notes/environments.md:152]: Keep in mind that we'll need to use this for TF, pulumi, and any IaC provider, so I'm hesitant to to adopt the bicep/arm format.

File: cli/azd/design-notes/environments.md [hunk 13]
```diff
@@ -0,0 +1,271 @@
...
+
+We create a new set of `azd infra config` commands with actions like `set`, `get`, `list`, `delete` to allow management of these parameters without having to edit the `config.json` file.
```
Comment (jongio) [inline cli/azd/design-notes/environments.md:154]: Consider a top level config command that manages the config file

File: cli/azd/design-notes/environments.md (hunk 10 above)
Comment (jongio) [inline cli/azd/design-notes/environments.md:107]: Let's fully flesh out what the current challenges are.  Is the challenge not being able to clearly articulate what an environment is?

An environment is metadata that describes an Azure application.

We could change the name to 'project' or 'solution' if that helps.

**Review by jongio — COMMENTED** [review]
Before we move forward with this I would like to:

1. Fully articulate what the challenges are and how these changes would help with each of those challenges
2. See a POC of what the new system would look like

File: cli/azd/design-notes/environments.md (hunk 13 above)
Comment (wbreza) [inline cli/azd/design-notes/environments.md:154]: In my opinion this area should be left up to the provision providers to implement how parameters are passed into the the respective provisioner.  In bicep things can stay the same - bicep will interpolate your template and replace with parameter values.

In terraform for example it can pass the parameters into the terraform plan using `--var` flags, etc.

File: cli/azd/design-notes/environments.md [hunk 14]
```diff
@@ -0,0 +1,271 @@
...
+@description('The Azure location to use for all resources.')
+@allowed([
+    'westus2'
+    'centralus'
+])
+@metadata({
+    azd: {
+        type: 'location'
+    }
+})
```
Comment (wbreza) [inline cli/azd/design-notes/environments.md:195]: I'm not familiar with this `metadata` field and we would need to ensure a similar concept is available on other provider implementations.  We may just need to fall back and use a naming convention that we look for for parameters across different provision providers.

File: cli/azd/design-notes/environments.md [hunk 15]
```diff
@@ -0,0 +1,271 @@
...
+```
+
+To start, we support the following types:
+
+`location`: The value should be the short name of an Azure location. All locations are supported by default, but the exiting `@allowed` may restrict this set.
+`azd-deployment-context`: The value should be an `azd` deployment context object. This is a JSON object that contains information about the project which may be useful for values when deploying. It will be described in further detail later.
+
+### Deployment Outputs
+
+When a deployment completes, the outputs from the deployment are stored in the `config.json` file under the "outputs" key (a sibling of "properties"). As we did on inputs, we re-use the schema of an outputs property of an ARM Deployment to save ourselves the trouble of inventing a new format.  When supporting Pulumi and Terraform we'll have to build an adapter but the format should be flexible enough to support this.
```
Comment (wbreza) [inline cli/azd/design-notes/environments.md:206]: What additional info would be in the `config.json` that isn't there today?

File: cli/azd/design-notes/environments.md [hunk 16]
```diff
@@ -0,0 +1,271 @@
...
+
+### Environment Files
+
+An explicit design goal is that all environment state is stored in single JSON file, which should provide flexibility in developing some longer term story around collaboration.
+
+While no longer technically required, we continue produce a `.env` and `.parameters.json` in the `.azure\<environment-name>` folder use by external tooling. These are generated artifacts that are updated any time the authoritative values are updated in `config.json`.
+
```
Comment (wbreza) [inline cli/azd/design-notes/environments.md:213]: Just calling out that the `*.parameters.json` files are just a side effect of the specific provision provider and would not need to exist for all providers like Pulumi or Terraform.

File: cli/azd/design-notes/environments.md [hunk 17]
```diff
@@ -0,0 +1,271 @@
...
+  "infra": {
+    "main": { 
+      "target": {
+        "type": "arm",
+        "subscriptionId": "faa080af-c1d8-40ad-9cce-e1a450ca5b57",
+        "location": "westus2"
+      }
+    }
+  }
+}
```
Comment (wbreza) [inline cli/azd/design-notes/environments.md:237]: Again - this is an implementation detail of our Bicep provider and we could choose to change this to also do RG deployments or make it configurable by the developer.

File: cli/azd/design-notes/environments.md [hunk 18]
```diff
@@ -0,0 +1,271 @@
...
+
+Note that the target **does not** contain information about a deployment name. Today the name we use for the deployment matches the environment name. We will change this to instead include the *project name* (stored in the name property of `azure.yaml` and the *environment id* to ensure uniqueness. We should also include some per deployment entropy (or perhaps just a timestamp) such that each deployment gets its own deployment object. This should help with some problems we face today where re-using the same deployment object can confuse our status reporting logic, and provides a better view of deployment history.  
+
+Since deployment names are limited to 64 characters, we will need someway to truncate names if we hit this limit.  We should bias towards keeping as much entropy as possible in the strings to ensure uniqueness at a given scope. We should tag these deployment objects with the full environment ID for querying later.
```
Comment (wbreza) [inline cli/azd/design-notes/environments.md:246]: Not quite sure we need commands to manage targets.  What we do need is a way to potentially split out how locations are handled with regard to environments.

Maybe this surfaces as a new `environments` section in the `azure.yaml` where we can specify something like `dev = [westus]`, `test = [eastus,westus]`, `prod = [eastus, centralus, westus]`.

I could also see scenarios where there are project services that should only be deployed in certain environments.  Like a dev or prod only service and have a way to call that out the `azure.yaml`.

File: cli/azd/design-notes/environments.md [hunk 19]
```diff
@@ -0,0 +1,271 @@
...
+
+## User Impacting Changes
+
+- We no longer add `AZURE_ENV_NAME`, `AZURE_LOCATION`, `AZURE_SUBSCRIPTION_ID` or `AZURE_PRINCIPAL_ID` to the `.env` file.
+- We no longer support environment variable substitution in the `.parameters.json` file in the `infra` folder, it is used just to configure static values.
+- In our templates we remove the `.parameters.json` files since they are no longer used (all values are user configured during first deployment). (*note: We can choose to keep these files if we think their existence provides value in our templates as a learning tool*)
+- We prompt and validate deployment parameters using the metadata applied in the bicep template.
+
+We can figure out how to rolls these changes out in a smooth enough way (for example, we could take the existence of a environment reference like `${FOO}` in a `.parameters.json` to mean we were on the "old" plan and `azd` can warn and do something sensible (or point folks at an article on how to upgrade).
```
Comment (wbreza) [inline cli/azd/design-notes/environments.md:255]: It was not 100% clear to me how this proposal eliminates the need for these variables/files.

**Review by wbreza — COMMENTED** [review]
Lots of information in here to understand & digest.  I'd recommend we all get together (possibly in person) and do like a 1/2 day or full day deep dive to understand your proposal here as well as gather feedback from the team.

File: cli/azd/design-notes/environments.md (hunk 16 above)
Comment (ellismg) [reply cli/azd/design-notes/environments.md:213]: Correct - I do think that when we bring terraform support online, we will write the "logical equivalent" of this file, which for terraform would be a tfvars file.  Note that bicep is exploring a bicep specific file that looks more like tfvars and so when that exists, I imagine we would start to produce that file format when your IaC of choice was bicep.

File: cli/azd/design-notes/environments.md (hunk 14 above)
Comment (ellismg) [reply cli/azd/design-notes/environments.md:195]: > We may just need to fall back and use a naming convention that we look for for parameters across different provision providers.

Alternately you could imagine encoding this metadata in JSON and storing it in an azd specific file side by side with whatever module defined the parameters and `azd` could use that file (if it exists).

File: cli/azd/design-notes/environments.md (hunk 11 above)
[the remark this answers is not in the document]
Comment (ellismg) [reply cli/azd/design-notes/environments.md:117]: > What is the value of the UUID? What is the value of the UUID? Would we use this for tagging purposes? or some other idea?

Yes, we'd use it for tagging purposes.  When we need to run resource queries to discover physical resources it provides a unique value to use instead of the environment name.  The value is just a randomly generated UUID to ensure uniqueness.

File: cli/azd/design-notes/environments.md (hunk 10 above)
Comment (karolz-ms) [reply cli/azd/design-notes/environments.md:107]: Well, there is no such thing as "Azure application" today, only applications that can be deployed to Azure. 

Our environment holds metadata about an Azure deployment of an application. It contains both "input" data, that is, data that is necessary for AZD CLI to deploy the application, as well as "output" data, that is, data that is a result of a deployment. This "input vs output" difference, and the fact that the user can add extra data to the environment, data that is neither necessary for deployment, nor a result of a deployment, plays a role in making the environment a bit confusing.

"Project" and "solution" have a very specific meaning in the dev tooling world, and calling environments either "project" or "solution" will increase the confusion quite a bit, let's not do that please.

File: cli/azd/design-notes/environments.md (hunk 11 above)
[the remark this answers is not in the document]
Comment (karolz-ms) [reply cli/azd/design-notes/environments.md:117]: UUIDs are not very user-friendly. I would try hard to find a way to use a tag that is human readable and chosen by the user before falling back to a pure UUID. Or at least have some human-readable prefix for the tag.

File: cli/azd/design-notes/environments.md [hunk 20]
```diff
@@ -0,0 +1,271 @@
...
+
+When `azd provision` is runs it loads the project wide configuration from the corresponding `.parameters.json` file in the infrastructure folder and merges with environment specific deployment parameters. If any required deployment parameters are missing the user is prompted for values and the responses are stored in the environment for future deployments.
+
+The deployment parameters are stored in `config.json` under the `infra` section, segmented by module.
+
+```json
+{
+  "id": "9fdfa455-2131-4faf-b2e3-8380aba97c96",
+  "name": "default",
+  "infra": {
```
Comment (karolz-ms) [inline cli/azd/design-notes/environments.md:142]:
```suggestion
  "deploymentParameters": {
```
Everything in this file is about infrastructure, so naming the deployment parameters section "infra" is not super intuitive

File: cli/azd/design-notes/environments.md [hunk 21]
```diff
@@ -0,0 +1,271 @@
...
+        type: 'location'
+    }
+})
+param location string
+```
+
+To start, we support the following types:
+
+`location`: The value should be the short name of an Azure location. All locations are supported by default, but the exiting `@allowed` may restrict this set.
+`azd-deployment-context`: The value should be an `azd` deployment context object. This is a JSON object that contains information about the project which may be useful for values when deploying. It will be described in further detail later.
```
Comment (karolz-ms) [inline cli/azd/design-notes/environments.md:202]: The value of `azd-deployment-context` is not clear to me. Why have a "structure" instead of just individual properties like subscription ID, or whatever else is necessary for the deployment to be started?

Why is `location` special and not part of the `azd-deployment-context`?

Perhaps I do not understand the proposal well enough...

File: cli/azd/design-notes/environments.md (hunk 15 above)
Comment (karolz-ms) [inline cli/azd/design-notes/environments.md:206]: I like the separation of deployment inputs vs deployment outputs. That said, squeezing out deployment outputs out of a JSON file is significantly harder than working with .env file, both for the application, and for the tooling.

File: cli/azd/design-notes/environments.md [hunk 22]
```diff
@@ -0,0 +1,271 @@
...
+
+### Environment Files
+
+An explicit design goal is that all environment state is stored in single JSON file, which should provide flexibility in developing some longer term story around collaboration.
+
+While no longer technically required, we continue produce a `.env` and `.parameters.json` in the `.azure\<environment-name>` folder use by external tooling. These are generated artifacts that are updated any time the authoritative values are updated in `config.json`.
+
+The generated `.env` file no longer contains deployment parameters.
+
+Now it is just the union of outputs from the deployment and any keys managed manually via `azd env set`. Since this file is regenerated each time a deployment happens, if a user removes an output from their `.bicep` file, the corresponding output will be removed from the `.env` file.
```
Comment (karolz-ms) [inline cli/azd/design-notes/environments.md:216]: I assume the additional settings that are neither deployment parameters, nor deployment outputs, would go into its own section in the `config.json` file?

File: cli/azd/design-notes/environments.md [hunk 23]
```diff
@@ -0,0 +1,271 @@
...
+
+For now, we'll introduce a `target` property in the environment:
+
+```json
+{
+  "id": "9fdfa455-2131-4faf-b2e3-8380aba97c96",
+  "name": "default",
+  "infra": {
+    "main": { 
+      "target": {
```
Comment (karolz-ms) [inline cli/azd/design-notes/environments.md:230]: Why is target information under the main module? Or under "infra"/ deployment parameters section, for that matter? Doesn't it apply to the whole environment?

File: cli/azd/design-notes/environments.md [hunk 24]
```diff
@@ -0,0 +1,271 @@
...
+        "type": "arm",
```
Comment (karolz-ms) [inline cli/azd/design-notes/environments.md:231]: Instead of "discriminating union" with "type" property, consider having separate objects for ARM, Terraform, Pulumi etc

**Comment by puicchan** [comment]
@ellismg, as you can see from UX Lab report, environment name is confusing to many developers. The current design will help to clarify. So I suggest to separate this out into two: Current Design and Proposed Model. The latter will require more discussion but the former will be beneficial to our customers and good to share.

File: cli/azd/design-notes/environments.md [hunk 25]
```diff
@@ -0,0 +1,271 @@
...
+
+As a UX concern, we may need to handle the "location" parameter in a clever way to reuse whatever location is being set for the template. Ideally we will not have to ask separate questions for "where do you want to store your resources" and "where do you want to store your deployment metadata" even though these are individually configurable.
+
+Note that the target **does not** contain information about a deployment name. Today the name we use for the deployment matches the environment name. We will change this to instead include the *project name* (stored in the name property of `azure.yaml` and the *environment id* to ensure uniqueness. We should also include some per deployment entropy (or perhaps just a timestamp) such that each deployment gets its own deployment object. This should help with some problems we face today where re-using the same deployment object can confuse our status reporting logic, and provides a better view of deployment history.
```
Comment (karolz-ms) [inline cli/azd/design-notes/environments.md:244]: I suspect some users might want to control what goes into the deployment name. The design should allow for that, even if we do not make it easy via CLI initially

File: cli/azd/design-notes/environments.md (hunk 5 above)
Comment (karolz-ms) [reply cli/azd/design-notes/environments.md:269]: It is also used by the tooling, big VS specifically, in order to determine what has been already provision, and whether there is a need to re-provision.

Comment (karolz-ms) [reply cli/azd/design-notes/environments.md:269]: Take a look at what `env refresh --output json` shows

File: cli/azd/design-notes/environments.md [hunk 26]
```diff
@@ -0,0 +1,271 @@
...
+
+Critically this means that every operation in `azd` requires interacting with an environment, and this it ends up being the first concept introduced when using `azd`.
+
+During environment creation the user is prompted for the following pieces of information:
+
+1. A Name for the environment.
+2. An Azure location.
+3. An Azure subscription.
+
+Unlike an Azure location or subscription, an environment is a new idea that has no mapping to any existing Azure concepts.  We don't give a clear explanation of how you should pick and environment name or how that name influences the rest of the tool.  We simply ask for an environment name and then ask again if you given us an "invalid" name.
```
[the remark this answers is not in the document]
Comment (weikanglim) [reply cli/azd/design-notes/environments.md:26]: I think so too. Most developers that work on services are familiar with the concept of environments. It's how they manage dev/test/prod, and also potentially how they manage regional deployments.

File: cli/azd/design-notes/environments.md (hunk 10 above)
Comment (weikanglim) [reply cli/azd/design-notes/environments.md:107]: I'm still not convinced that `azd` is introducing something novel -- environment as a concept is an industry-wide term used for developers that work on services.

The definition of an environment for me:

A collection of resources that form an instance of the application service. This can be used to isolate non-production vs production instances, or to create regional instances when necessary.

And I don't think `azd` is doing something different ahn that.

Comment (karolz-ms) [reply cli/azd/design-notes/environments.md:107]: @weikanglim respectfully disagree. In your definition ("a collection of resources that form an instance of the application service") I assume you mean "a container". If that is the case, you are describing a cloud-side, or server-side concept: a resource group, datacenter, Kubernetes cluster etc. 

This is different from azd environment, which (today) is a client-side concept with no cloud-side counterpart. It is just a named set of metadata related to application deployment.

Comment (weikanglim) [reply cli/azd/design-notes/environments.md:107]: @karolz-ms 

> In your definition ("a collection of resources that form an instance of the application service") I assume you mean "a container". If that is the case, you are describing a cloud-side, or server-side concept: a resource group, datacenter, Kubernetes cluster etc.

I think we mean the same, but just to clarify -- "A collection of resources" here is more abstract than server-side groupings. It simply states that we have a set of resources that when formed together, becomes the application that is being delivered. It can be multiple resource groups, multiple datacenters or clusters, or perhaps just a single VM.

> This is different from azd environment, which (today) is a client-side concept with no cloud-side counterpart. It is just a named set of metadata related to application deployment.

Yes, I was not being technically correct here. I was focused on the actual usage of the metadata because that's what customers use `azd` environments to do, rather than describing what an `azd environment` is in terms of what it holds. Let me attempt at redefining it:

An environment describes a collection of resources that form the application and can be deployed. Typical environments include: development, test, staging or production.

I think this is very similar to how Azure DevOps define their environments, which is also metadata that ADO stores on their service that facilitates multi-stage deployments: https://docs.microsoft.com/en-us/azure/devops/pipelines/process/environments?view=azure-devops

Comment (karolz-ms) [reply cli/azd/design-notes/environments.md:107]: I guess it is useful to differentiate between what an environment IS vs what it is USED FOR. From the latter perspective it is used to provision/deploy separate instances of an application, possibly on a different scale, different access permissions, location etc. etc. (although today `azd` does not help you much with all these aspects). I think this is what you are describing and I agree we need to include this information when we talk about AZD environments.

It is also worth noting that today `azd` environment can really only manage a single application instance. This is different from AzDO environments or, say, Azure Container Applications environments.
